File: Pictok/Game/ArchiveSession.swift
import Foundation
import Observation

/// Single-puzzle game session pinned to one past Daily puzzle, used by the
/// Archive feature. Mirrors EndlessSession in shape but never advances and
/// records its outcome via `UserStateStore.recordArchiveOutcome` (which is
/// streak-neutral by design).
@Observable
final class ArchiveSession {
    let puzzle: Puzzle
    private let store: UserStateStore

    private(set) var hearts: Int = 5
    private(set) var correctGuesses: Set<Character> = []
    private(set) var wrongGuesses: Set<Character> = []
    private(set) var isSolved: Bool = false
    private(set) var isFailed: Bool = false
    private(set) var hintUsedThisPuzzle: Bool = false
    private(set) var hasShownOneChanceWarning: Bool = false

    init(puzzle: Puzzle, store: UserStateStore) {
        self.puzzle = puzzle
        self.store = store
    }

    var needsSubmit: Bool {
        guard !isSolved, !isFailed else { return false }
        return GameEngine.isSolved(answer: puzzle.answer,
                                   correctGuesses: correctGuesses,
                                   revealedLetter: nil)
    }

    func guess(letter: Character) {
        guard !isSolved, !isFailed else { return }
        let upper = Character(String(letter).uppercased())
        guard !correctGuesses.contains(upper), !wrongGuesses.contains(upper) else { return }

        if GameEngine.isCorrect(letter: upper, in: puzzle) {
            correctGuesses.insert(upper)
        } else {
            wrongGuesses.insert(upper)
            hearts -= 1
            if hearts == 1 && !hasShownOneChanceWarning {
                hasShownOneChanceWarning = true
            }
            if GameEngine.isFailed(lives: hearts) {
                isFailed = true
                recordOutcome(solved: false)
            }
        }
    }

    func submit() {
        guard needsSubmit else { return }
        isSolved = true
        recordOutcome(solved: true)
    }

    func useHint() {
        guard !hintUsedThisPuzzle, !isSolved, !isFailed else { return }
        guard let letter = GameEngine.letterToReveal(for: puzzle, correctGuesses: correctGuesses) else { return }
        correctGuesses.insert(letter)
        hintUsedThisPuzzle = true
    }

    private func recordOutcome(solved: Bool) {
        store.recordArchiveOutcome(
            puzzleId: puzzle.id,
            solved: solved,
            wrongGuesses: wrongGuesses.count,
            hintUsed: hintUsedThisPuzzle,
            date: puzzle.date
        )
    }
}

File: web/js/archive-session.js
import * as engine from './game-engine.js';
import * as us from './user-state.js';

const MAX_HEARTS = 5;

/**
 * Creates a session pinned to a single past Daily puzzle. Records its outcome
 * via `us.recordArchiveOutcome` (streak-neutral) on solve or fail. Mirrors
 * iOS `ArchiveSession`. The optional `storage` argument is forwarded to
 * `us.save` after each terminal write; pass `globalThis.localStorage` in
 * production, omit for unit tests that only inspect the in-memory state.
 */
export function createArchiveSession(puzzle, state, storage) {
  const session = {
    puzzle,
    hearts: MAX_HEARTS,
    correct: new Set(),
    wrong: new Set(),
    solved: false,
    failed: false,
    hintUsed: false,
    hasShownOneChanceWarning: false,

    get needsSubmit() {
      if (session.solved || session.failed) return false;
      return engine.isSolved(puzzle.answer, session.correct, null);
    },

    guess(letter) {
      if (session.solved || session.failed) return;
      const u = String(letter).toUpperCase();
      if (session.correct.has(u) || session.wrong.has(u)) return;

      if (engine.isCorrect(u, puzzle.answer)) {
        session.correct.add(u);
      } else {
        session.wrong.add(u);
        session.hearts -= 1;
        if (session.hearts === 1 && !session.hasShownOneChanceWarning) {
          session.hasShownOneChanceWarning = true;
        }
        if (engine.isFailed(session.hearts)) {
          session.failed = true;
          recordOutcome(state, session, false, storage);
        }
      }
    },

    submit() {
      if (!session.needsSubmit) return;
      session.solved = true;
      recordOutcome(state, session, true, storage);
    },

    useHint() {
      if (session.hintUsed || session.solved || session.failed) return;
      const letter = engine.letterToReveal(puzzle.answer, session.correct);
      if (!letter) return;
      session.correct.add(letter);
      session.hintUsed = true;
    },
  };

  return session;
}

function recordOutcome(state, session, solved, storage) {
  us.recordArchiveOutcome(state, session.puzzle, {
    solved,
    wrongGuesses: session.wrong.size,
    hintUsed: session.hintUsed,
  });
  if (storage) us.save(state, storage);
}
